internal/policy/filter.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2020 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policy
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// RegexFilter represents a regular expression filter
+type RegexFilter struct {
+	filtered map[string]string
+
+	Regexp  *regexp.Regexp
+	Replace string
+}
+
+// NewRegexFilter constructs new RegexFilter object
+func NewRegexFilter(pattern string, replace string) (*RegexFilter, error) {
+	m, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("invalid regular expression pattern '%s': %s", pattern, err.Error())
+	}
+	return &RegexFilter{
+		Regexp:  m,
+		Replace: replace,
+	}, nil
+}
+
+// Apply will construct the filtered list of tags based on the provided list of tags
+func (f *RegexFilter) Apply(list []string) {
+	f.filtered = map[string]string{}
+	for _, item := range list {
+		if f.Regexp.MatchString(item) {
+			tag := item
+			if f.Replace != "" {
+				tag = f.Regexp.ReplaceAllString(item, f.Replace)
```

Comment:
I think you can do this without repeating work, using Find*Index and Expand.

Reply:
Not sure I follow here, what do you refer to as repeated work? I would assume it'll look kind of like this, right?
```go
if submatches := f.Regexp.FindStringSubmatchIndex(item); len(submatches) > 0 {
	tag := item
	if f.Replace != "" {
		result := []byte{}
		result = f.Regexp.ExpandString(result, f.Replace, item, submatches)
		tag = string(result)
	}
	f.filtered[tag] = item
}
```

The above works just fine but I'm not sure I see the benefit :confused:

Reply:
Both `MatchString` and `ReplaceAllString` "run" the regular expression -- it's just that the former disregards submatches, and the latter _also_ does substitution. On the other hand, `FindStringSubmatchIndex` runs the regular expression to get the submatches (and whether it matches at all), but ExpandString does only the substitution work.

It is unlikely to make a vast difference to any individual match, but all else being equal, it's worth avoiding extra calculation.

Reply:
Got it, thanks for the explanation, I just looked into more detail in the implementation of `ReplaceAllString` vs `ExpandString` and I did notice now what you mean. I will refactor.